### PR TITLE
Heterogeneous Treatment Effects

### PR DESCRIPTION
**Pull request by frederickluser**
Dear Grant,

I implemented now the _heterogeneous treatment effects_ I mentioned in issue #13.

Interacting the treatment with a time-constant covariate allows estimating the effect separately for multiple groups, for example, women and men (see the image below).

I use this now in my research, and maybe you find it interesting to include in your package.

All the best, Frederic

es

### CODE REVIEW

File: R/emfx.R
Comment (grantmcdermott) [inline R/emfx.R:28]: Undeclared stringr dependency. Please use base `grepl` instead.

File: R/emfx.R
Comment (grantmcdermott) [inline R/emfx.R:66]: Seeing the `[1]` index requirement makes me a little nervous that we may be introducing unwanted changes to `by_var`. Can you tell me why it's necessary here? (Or, alternatively, what happens if we don't use it and leave the code as it was?)

File: R/emfx.R
Comment (grantmcdermott) [inline R/emfx.R:53]: Why is this necessary?

File: R/emfx.R
Comment (grantmcdermott) [inline R/emfx.R:57]: Minor thing, but I'd prefer to use `=` for assignment instead of `<-` to be consistent.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:81]: Can we use a different set of argument names? For consistency with the other arguments, I'd prefer something like `xvar` and `xref`.

Also: You need to document the arguments as part of the roxygen2 preamble, otherwise it's going to generate a bunch of CRAN check fails. See the top of the doc for examples.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:180]: Please change this back to the original formulation. I don't see what switching out the idiomatic `i.` fixest syntax for `as_factor` is giving us besides another undeclared tidyverse dependency.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:238]: As per my earlier comment, please use the idiomatic `i(gvar, i.tvar)` syntax instead of `as_factor`. You may need to bring intvar (xvar) to the front, so that it's written as `intvar:i(gvar, i.tvar)`.  Also: what's happening to `gref`, `tref`, etc. here? Have they dropped out by mistake?

You'll ultimately need to test with `devtools::check` or `tinytest::run_build_dir("inst/tinytest")`.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:231]: Another undeclared (in the DESCRIPTION file) tidyverse dependency . Realistically, I'm not going to bring in the whole of dplyr (and all its dependencies) for a single line of code that could be done pretty easily with base R (or fixest::demean).  Stepping back, can you say a bit about why this is using a different demeaning approach than the group-level demeaning approach we use for the regular covariates? Apologies if I have misunderstood something.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:238]: See earlier comments about `as_factor`.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:268]: I don't think we need the if...else condition here. If intvar (xvar) is NULL then it's still fine to include it in the return list since it won't affect the downstream methods.

File: R/etwfe.R
Comment (grantmcdermott) [inline R/etwfe.R:113]: I might be misunderstanding what's happening here, but is there a reason that the id (xref) cannot be guessed at from the data, like we do for `gref` and `tref`?

**Review by grantmcdermott — CHANGES_REQUESTED** [review]
Thanks Frederick, this looks promising. There are a bunch of things we'll need to fix before I can consider the merge. I've left quite a few comments, so please take a look and then run `devtools::check()` locally to see if you've caught the stuff that's going to cause CRAN check failures.

**Comment by frederickluser** [comment]
Dear Grant,

I implemented all your requested changes and ran the check function locally on my computer.
I received no errors, warnings, or notes.

Additionally, I added the option to run a hypothesis test for the heterogeneous effects in the `emfx` file. Maybe you like this option.

**Comment by grantmcdermott** [comment]
Super, thanks Frederik. I appreciate you bearing with me and think we're close now. I have three more requests and then will merge after a final check:

- Could you please add a test to `inst/`tinytest`. I'm fine for you to use simulated data if that's easiest. Ideally, however, I want to make sure that we're benchmarking against a known set of results, since I obviously don't want to push through new changes that might give misleading results to users.
- Similarly, if you know of a good example dataset that we can use to demonstrate the `xvar` functionality---e.g., in the vignette---please do say.
- Please add yourself as a contributor in the DESCRIPTION file.

LMK if anything is unclear, or if you'd like to discuss any of these requests. Thanks!

**Comment by frederickluser** [comment]
Thank you for your feedback, Grant. Regarding your points:

- I implemented the `tinytest`. I called them `test_emfx_xvar.R` and `test_etwfe_xvar.R` and used in both the `mpdta` dataset so far. 
To test the `xvar` argument, I added a random binary or categorical variable (`xvar` also works with more than two values, while continuous variable make less sense. Nonetheless, the code also runs for continuous variables).
Finally, I ran again a `devtools::check` that looks fine.

- I could not find so far a nice dataset. I tested so far the `mpdta` as well as `data("pj_officer_level_balanced", package = "staggered")` (large, with 500k observations) as well as simulated data and the results are as expected. 
I also ran the function against the `did` package and a standard TWFE. I attach my file where I do these comparisons:
[etwfe_comparisons.zip](https://github.com/grantmcdermott/etwfe/files/10490949/etwfe_comparisons.zip)

- I added myself in the description.

Final point - I found a way to get rid of the `[1]` index in _emfx_. I think it's less error-prone like this.

**Comment by grantmcdermott** [comment]
This is great. I have a couple tweaks that I'd like to make on the backend, but happy to merge as-is. Many thanks for the nice contribution!

**Comment by grantmcdermott** [comment]
HU that `emfx` now autodetects whether to estimate het. TEs based on whether `xvar` is detected as an attribute of the underlying emfx object. Users can override this behavior with the new "by_xvar" argument. See #21.

**Comment by frederickluser** [comment]
very nice!

**Comment by frederickluser** [comment]
A related thought - if the user enters a continuous xvar (which is fine), marginaleffects should evaluate at a meaningful value of xvar, e.g., the mean. 

You think this might be something so think about?

**Comment by grantmcdermott** [comment]
> A related thought - if the user enters a continuous xvar (which is fine), marginaleffects should evaluate at a meaningful value of xvar, e.g., the mean.

Sounds like a good idea. But I'm not sure I'll have the bandwidth to incorporate before the next CRAN submission. I'm hoping to do today or tomorrow, since the current version is error'ing out as a result of the upstream marginaleffects 0.9.0 changes (fixed in this dev version). But please feel free to raise another issue/PR if you have an implementation plan.

**Pull request merged**
